### PR TITLE
feat: GPU support

### PR DESCRIPTION
**Pull request by alexlokshin-czi**
This PR adds the ability to schedule services on GPU nodes, and specify GPU limits on them. For GPU-intense workloads to function, EKS cluster needs GPU nodes, `k8s-device-plugin` (https://github.com/NVIDIA/k8s-device-plugin) needs to be installed, and the service has to have `platform_architecture = "amd64"` and have the following selector present:

```HCL
 additional_node_selectors = { "nvidia.com/gpu.present" = "true" }
 ```

<details open>
  <summary><a href="https://czi-tech.atlassian.net/browse/CCIE-1747" title="CCIE-1747" target="_blank">CCIE-1747</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
  <td>Add GPU nodepool support for happy</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Task
        Task
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Review</td>
    </tr>
  </table>
</details>

---

Addresses [CCIE-1747].

[CCIE-1747]: https://czi-tech.atlassian.net/browse/CCIE-1747?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: examples/gpu_app/.happy/terraform/envs/rdev/main.tf [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      path = "/*"
+      // the platform architecture of the container. this should match what is in 
+      // the platform attribute of your docker-compose.yml file for your service.
+      // oneof: amd64, arm64.
+      // Try to always select arm since it comes with a lot of cost savings and performance
+      // benefits and has little to no impact on developers.
+      platform_architecture = "arm64"
+
+      // Schedule to run on a GPU node
+      additional_node_selectors = {
```
Comment (alexlokshin-czi) [inline examples/gpu_app/.happy/terraform/envs/rdev/main.tf:48]: To make this work, you need GPU nodes in the cluster, k8s-device-plugin installed (will add to k8s-core if everything works), this selector, and a gpu limit.

File: examples/gpu_app/.happy/config.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                "memory": "128Mi",
+                "cpu": "100m",
+                "health_check_path": "/",
+                "service_type": "EXTERNAL",
+                "path": "/*",
+                "priority": 0,
+                "success_codes": "200-499",
+                "initial_delay_seconds": 30,
+                "period_seconds": 3,
+                "platform_architecture": "arm64",
```
Comment (gulbrand-czi) [inline examples/gpu_app/.happy/config.json:32]: I think this needs to be amd64

File: examples/gpu_app/.happy/terraform/envs/rdev/main.tf [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      // IMAGE_TEMPLATE: don't deploy any services, just use to create and push images
+      service_type = "INTERNAL"
+      // the path to reach this search
+      path = "/*"
+      // the platform architecture of the container. this should match what is in 
+      // the platform attribute of your docker-compose.yml file for your service.
+      // oneof: amd64, arm64.
+      // Try to always select arm since it comes with a lot of cost savings and performance
+      // benefits and has little to no impact on developers.
+      platform_architecture = "arm64"
```
Comment (gulbrand-czi) [inline examples/gpu_app/.happy/terraform/envs/rdev/main.tf:45]: amd64?

File: examples/gpu_app/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+version: "3.8"
+
+services:
+  frontend:
+    image: "frontend"
+    profiles: [ "*" ]
+    platform: linux/arm64
```
Comment (gulbrand-czi) [inline examples/gpu_app/docker-compose.yml:7]: amd64?

**Review by gulbrand-czi — COMMENTED** [review]
I think we need to use amd64 not arm64. If the GPU instances do actually support arm64, then that would be ideal, but I will have to see if I can compile my service to arm64 with all of the nvidia libs as well. I tried that and that did not work, but I can try again.

**Review by gulbrand-czi — APPROVED** [review]

**Comment by alexlokshin-czi** [comment]
> I think we need to use amd64 not arm64. If the GPU instances do actually support arm64, then that would be ideal, but I will have to see if I can compile my service to arm64 with all of the nvidia libs as well. I tried that and that did not work, but I can try again.

Yes, that was a mistake on my part. Updated.

File: terraform/modules/happy-stack-eks/variables.tf [hunk 5]
```diff
@@ -77,8 +78,9 @@ variable "services" {
     synthetics : optional(bool, false),
     initial_delay_seconds : optional(number, 30),
     period_seconds : optional(number, 3),
-    platform_architecture : optional(string, "amd64"), // Supported values: amd64, arm64
-    bypasses : optional(map(object({                   // Only used for INTERNAL service_type
+    platform_architecture : optional(string, "amd64"),     // Supported values: amd64, arm64; GPU nodes are amd64 only.
+    additional_node_selectors : optional(map(string), {}), // For GPU use: { "nvidia.com/gpu.present" = "true" }
```
Comment (jakeyheath) [inline terraform/modules/happy-stack-eks/variables.tf:82]: I think having this generally. Maybe in addition, should we add the nvidia specific node selectors if gpu is not null so you don't have to know to add both? Maybe also override the platform_architecture if gpu is not null

Comment (alexlokshin-czi) [reply terraform/modules/happy-stack-eks/variables.tf:82]: Done!

**Review by jakeyheath — APPROVED** [review]

**Pull request merged**